### PR TITLE
ENH: cproperties for datetime objects

### PR DESCRIPTION
**Pull request by jbrockmendel**
Open to suggestions for how to test these.

### CODE REVIEW

**Comment by scoder** [comment]
Tests should go into one of the `datetime*.pyx` tests. You can use a [tree assertion](https://github.com/cython/cython/wiki/HackerGuide#parse-tree-assertions) to test that there are no `AttributeNode`s left after resolving the properties into function calls at compile time, e.g. for `return (date.year, date.month, date.day)`. For non-trivial path expressions, it's usually a good idea to assert that something is there but something more specific isn't, to make sure the path expression itself is sane. But if you can get away with `//AttributeNode` not being there, that's simpe enough.

File: Cython/Includes/cpython/datetime.pxd [hunk 1]
```diff
@@ -24,7 +24,33 @@ cdef extern from "datetime.h":
         pass
 
     ctypedef extern class datetime.datetime[object PyDateTime_DateTime]:
-        pass
+        @property
+        cdef int year(self):
```
Comment (scoder) [inline Cython/Includes/cpython/datetime.pxd:28]: This needs to be declared `inline`, because that's the intention after all.

**Comment by scoder** [comment]
The test failure is inherited from current master. You can ignore it.
Tests for this would be nice, though. :)

**Comment by jbrockmendel** [comment]
> Tests for this would be nice, though. :)

I guess the way to test this would be to write code that uses a datetime object and check that the generated C code uses `PyDateTime_GET_WHATEVER` instead of `PyGetAttribute`?

**Comment by da-woods** [comment]
> > Tests for this would be nice, though. :)
> 
> I guess the way to test this would be to write code that uses a datetime object and check that the generated C code uses `PyDateTime_GET_WHATEVER` instead of `PyGetAttribute`?

[Parse tree assertions](https://github.com/cython/cython/wiki/HackerGuide#parse-tree-assertions) would probably be the way to do it instead of inspecting the C source code. I suspect doing:

```
cdef int year = somedatetime.year
```

and looking for the absence of a `CoerceFromPyTypeNode` is probably the easiest way to spot that it's using the properties directly.

**Comment by jbrockmendel** [comment]
Looks like Travis failure is just a build timeout?

**Comment by da-woods** [comment]
> Looks like Travis failure is just a build timeout?

I think something's crashing and it's hard to see that across the multiprocessing boundary (so it keeps waiting for that process to finish). However, it's nothing to do with this PR.

**Comment by scoder** [comment]
> I think something's crashing

That's the embedding test. I have yet to look into fixing that, or at least getting that out of the way. The failure is really annoying.

File: tests/run/datetime_pxd.pyx [hunk 2]
```diff
@@ -212,3 +210,20 @@ def test_timedelta_total_seconds():
...
+    """
+    cdef:
+        datetime dt = datetime(2020, 8, 18, 4, 9)
+
+    dt.year
+    dt.month
+    dt.day
+    dt.hour
+    dt.second
+    dt.microsecond
```
Comment (scoder) [inline tests/run/datetime_pxd.pyx:229]:
```suggestion
@cython.test_fail_if_path_exists(
    "//CoerceFromPyTypeNode",
    "//AttributeNode",
)
def test_datetime_attrs_inlined(datetime dt):
    # GH#3737
    """
    >>> py_dt = datetime(2020, 8, 18, 4, 9)
    >>> dt = test_datetime_attrs_inlined(py_dt)
    >>> dt[:5]
    (2020, 8, 18, 4, 9)
    >>> dt[5] == py_dt.second  or  (dt[5], py_dt.second)
    True
    >>> dt[6] == py_dt.microsecond  or  (dt[6], py_dt.microsecond)
    True
    """
    return (
        dt.year,
        dt.month,
        dt.day,
        dt.hour,
        dt.minute,
        dt.second,
        dt.microsecond,
    )
```

File: tests/run/datetime_pxd.pyx [hunk 3]
```diff
@@ -212,18 +212,28 @@ def test_timedelta_total_seconds():
...
-def test_datetime_attrs_inlined():
+@cython.test_fail_if_path_exists(
+    "//CoerceFromPyTypeNode",
+    "//AttributeNode",
+)
+def test_datetime_attrs_inlined(datetime dt):
     # GH#3737
     """
-    >>> test_datetime_attrs_inlined()
+    >>> py_dt = datetime(2020, 8, 18, 4, 9)
```
Comment (scoder) [inline tests/run/datetime_pxd.pyx:222]:
```suggestion
    >>> from datetime import datetime
    >>> py_dt = datetime(2020, 8, 18, 4, 9)
```

**Comment by scoder** [comment]
Thanks

**Pull request merged**